Specs for chat completion API + text-generation stream

Wauplin opened the pull request:
Based on OpenAI openapi specs: https://github.com/openai/openai-openapi/blob/master/openapi.yaml (MIT license) +some help from hf.co/chat :smile:

### What does this PR do:

- ~add Python generating script (i.e. work from https://github.com/huggingface/huggingface.js/pull/505 that should be merged first)~ => done
- adapt inference-codegen script to generate from `output_stream.json` as well
- add chat-completion specs: input.json, output.json and output_stream.json
- update text-generation output.json => changed some names, define `Details` and `SequenceDetails` separately even though they share some properties (I feel it's more readable), added some TGI-specific optional fields in output details 
- add output_stream.json for text-generation and support for `stream: bool` in input.json

### Differences compared to openai chat-completion specs:

In general my approach has been to take openai specs and remove fields that are not supported by us. Fields that are define by both openai and these specs are consistent. All fields defined in these specs exists in openai specs (i.e. no new fields compared to them). => This should be robust if we want to add new fields later / support more cases while not breaking consistency.

Removed in input.json:
- model
- presence_penalty
- tools
- n
- logit_bias
- logprobs
- top_logprobs
- FunctionMessage
- ToolMessage
- Optional names for UserMessage/AssistantMessage/SystemMessage

Removed in output.json:
- model
- id
- system_fingerprint
- object
- usage

Wauplin commented on packages/tasks/src/tasks/chat-completion/inference.ts:
```diff
@@ -0,0 +1,128 @@
+	 */
+	frequency_penalty?: number;
+	/**
+	 * The maximum number of tokens that can be generated in the chat completion. The total
+	 * length of input tokens and generated tokens is limited by the model's context length.
+	 */
+	max_tokens?: number;
+	messages?: MessageElement[];
+	/**
+	 * ID of the model to use. See the model endpoint compatibility table for details on which
+	 * models work with the Chat API.
+	 */
+	model?: string;
+	/**
+	 * The random sampling seed.
+	 */
+	seed?: number;
+	/**
+	 * Up to 4 sequences where the API will stop generating further tokens.
+	 */
+	stop?: string;
+	/**
+	 * If set, partial message deltas will be sent, like in ChatGPT. Tokens will be sent as
+	 * data-only server-sent events as they become available, with the stream terminated by a
+	 * data: [DONE] message.
+	 */
+	stream?: boolean;
+	/**
+	 * What sampling temperature to use, between 0 and 2. Higher values like 0.8 will make the
+	 * output more random, while lower values like 0.2 will make it more focused and
+	 * deterministic. We generally recommend altering this or top_p but not both.
+	 */
+	temperature?: number;
+	/**
+	 * An alternative to sampling with temperature, called nucleus sampling, where the model
+	 * considers the results of the tokens with top_p probability mass. So 0.1 means only the
+	 * tokens comprising the top 10% probability mass are considered. We generally recommend
+	 * altering this or temperature but not both.
+	 */
+	top_p?: number;
```
So far the only input parameter that have changed is `max_new_tokens` (in transformers/TGI) vs `max_tokens` (in openai specs). I can revert to `max_new_tokens` though.

> (it's ok if the API is not exactly the same imo no?)

I have mixed feelings about that TBH. I feel that we should stick to be as close as possible in the clients (even if it means having `max_new_tokens = max_tokens` somewhere in the code). And if/once we implement this API server-side, keep the openai-compatible spec.

Wauplin replied:
By the way, TGI already has support for chat completion API and it defines `max_tokens` (see [here](https://github.com/huggingface/text-generation-inference/blob/7dbaf9e9013060af52024ea1a8b361b107b50a69/router/src/lib.rs#L599)) so let's keep that.

Wauplin replied:
**Erratum:** we have `stop_sequences` and `repetition_penalty`. OpenAI has `stop` and `frequency_penalty`.

For `frequency_penalty`, I'm not sure the value means exactly the same :confused:

julien-c replied:
ping the authors of TGI about this! ^

SBrandeis commented on packages/tasks/src/tasks/chat-completion/spec/input.json:
```diff
@@ -0,0 +1,99 @@
+					"description": "The contents of the assistant message."
+				}
+			},
+			"required": ["role"]
```
Isn't `"content"` required for assistant messages as well?
```suggestion
			"required": ["role", "content"]
```

SBrandeis commented on packages/tasks/src/tasks/chat-completion/spec/input.json:
```diff
@@ -0,0 +1,99 @@
+				"role": {
+					"const": "system",
+					"description": "The role of the messages author, in this case system."
+				},
+				"content": {
+					"type": "string",
+					"description": "The contents of the system message."
+				}
+			},
+			"required": ["role", "content"]
+		},
+		"UserMessage": {
+			"type": "object",
+			"properties": {
+				"role": {
+					"const": "user",
+					"description": "The role of the messages author, in this case user."
+				},
+				"content": {
+					"type": "string",
+					"description": "The contents of the user message."
+				}
+			},
+			"required": ["role", "content"]
+		},
+		"AssistantMessage": {
+			"type": "object",
+			"properties": {
+				"role": {
+					"const": "assistant",
+					"description": "The role of the messages author, in this case assistant."
+				},
+				"content": {
+					"type": "string",
+					"description": "The contents of the assistant message."
+				}
+			},
+			"required": ["role"]
+		}
+	}
```
I think those can be grouped in a single definition

```suggestion
		"Message": {
			"type": "object",
			"properties": {
				"role": {
                    "oneOf": [
                        { "const": "assistant" },
                        { "const": "system" },
                        { "const": "user" },
                    ],
					"description": "The role of the messages author."
				},
				"content": {
					"type": "string",
					"description": "The contents of the message."
				}
			},
			"required": ["role", "content"]
		},
	}
```

SBrandeis commented on packages/tasks/src/tasks/chat-completion/spec/output.json:
```diff
@@ -0,0 +1,54 @@
+				"properties": {
+					"finish_reason": {
+						"type": "string",
+						"description": "The reason the model stopped generating tokens."
```
```suggestion
						"description": "The reason why the model stopped generating tokens."
```

SBrandeis commented on packages/tasks/src/tasks/chat-completion/spec/output.json:
```diff
@@ -0,0 +1,54 @@
+			"description": "This fingerprint represents the backend configuration that the model runs with."
+		}
+	},
+	"required": ["id", "choices", "created", "model", "system_fingerprint", "object", "usage"]
```
Those two properties are not defined
```suggestion
	"required": ["id", "choices", "created", "model", "system_fingerprint"]
```

SBrandeis commented on packages/tasks/src/tasks/chat-completion/spec/input.json:
```diff
@@ -0,0 +1,99 @@
+	"description": "Inputs for ChatCompletion inference",
+	"type": "object",
+	"properties": {
+		"messages": {
+			"type": "array",
+			"items": {
+				"oneOf": [
+					{
+						"$ref": "#/$defs/SystemMessage"
+					},
+					{
+						"$ref": "#/$defs/UserMessage"
+					},
+					{
+						"$ref": "#/$defs/AssistantMessage"
+					}
+				]
+			}
+		},
+		"model": {
+			"type": "string",
+			"description": "ID of the model to use. See the model endpoint compatibility table for details on which models work with the Chat API."
+		},
+		"frequency_penalty": {
```
Those two props should be required no? Maybe only the `messages`

SBrandeis commented on packages/tasks/src/tasks/chat-completion/inference.ts:
```diff
@@ -0,0 +1,128 @@
+	[property: string]: unknown;
+}
+
+export interface MessageElement {
+	/**
+	 * The contents of the system message.
+	 *
+	 * The contents of the user message.
+	 *
+	 * The contents of the assistant message.
+	 */
+	content?: string;
+	/**
+	 * The role of the messages author, in this case system.
+	 *
+	 * The role of the messages author, in this case user.
+	 *
+	 * The role of the messages author, in this case assistant.
+	 */
+	role: Role;
+	[property: string]: unknown;
+}
+
+export type Role = "system" | "user" | "assistant";
+
```
Quite amazed by the Quicktype code generation here :o

SBrandeis commented on packages/tasks/src/tasks/chat-completion/inference.ts:
```diff
@@ -0,0 +1,128 @@
+	 * data-only server-sent events as they become available, with the stream terminated by a
+	 * data: [DONE] message.
+	 */
+	stream?: boolean;
```
Am I right to say the streaming output is not defined/typed in the spec?

Wauplin replied:
Correct. I'll work on that as well.

SBrandeis reviewed (commented):
Nice, thank you for looking into this!

This API is quite different from the other tasks (which have `inputs` & `parameters` fields). Still, I think it can be beneficial to match OpenAI API given its popularity 👍

SBrandeis commented on packages/tasks/src/tasks/chat-completion/inference.ts:
```diff
@@ -0,0 +1,128 @@
+	 * The random sampling seed.
+	 */
+	seed?: number;
+	/**
+	 * Up to 4 sequences where the API will stop generating further tokens.
+	 */
+	stop?: string;
```
Would be worth specifying how this is "encoded" in the string (I guess it's comma-separated?)

SBrandeis commented on packages/tasks/scripts/inference-codegen.ts:
```diff
@@ -44,6 +54,12 @@ async function buildInputData(taskId: string, taskSpecDir: string, allSpecFiles:
 		name: `${taskId}-output`,
 		schema: await fs.readFile(`${taskSpecDir}/output.json`, { encoding: "utf-8" }),
 	});
+	if (taskId === "text-generation" || taskId === "chat-completion") {
+		await schema.addSource({
+			name: `${taskId}-stream-output`,
+			schema: await fs.readFile(`${taskSpecDir}/output_stream.json`, { encoding: "utf-8" }),
+		});
```
nice!!

SBrandeis commented on packages/tasks/src/tasks/text-generation/spec/output.json:
```diff
@@ -10,43 +10,42 @@
 	"$defs": {
+		"FinishReason": {
+			"type": "string",
+			"title": "TextGenerationFinishReason",
```
:ok_hand: Nice to add the `TextGeneration` prefix to intermediate types

SBrandeis commented on packages/tasks/src/tasks/text-generation/inference.ts:
```diff
@@ -114,62 +118,69 @@ export interface TextGenerationOutputDetails {
 	/**
 	 * The generated text
 	 */
-	generated_text: number;
```
lol good catch

SBrandeis commented on packages/tasks/src/tasks/text-generation/spec/output.json:
```diff
@@ -10,43 +10,42 @@
+			"title": "TextGenerationFinishReason",
+			"description": "The reason why the generation was stopped.",
+			"oneOf": [
+				{ "const": "length", "description": "The generated sequence reached the maximum allowed length" },
+				{ "const": "eos_token", "description": "The model generated an end-of-sentence (EOS) token" },
+				{ "const": "stop_sequence", "description": "One of the sequence in stop_sequences was generated" }
```
nit: maybe let's add the value in the description so the generated docstring is more helpful
```suggestion
				{ "const": "length", "description": "length: The generated sequence reached the maximum allowed length" },
				{ "const": "eos_token", "description": "eos_token: The model generated an end-of-sentence (EOS) token" },
				{ "const": "stop_sequence", "description": "stop_sequence: One of the sequence in stop_sequences was generated" }
```

Wauplin replied:
Definitely a good idea!

SBrandeis commented on packages/tasks/src/tasks/chat-completion/spec/input.json:
```diff
@@ -0,0 +1,54 @@
+				"type": "object",
+				"properties": {
+					"role": {
+						"oneOf": [{ "const": "assistant" }, { "const": "system" }, { "const": "user" }],
```
(nit) Maybe add a title here too to differentiate the type?

```suggestion
						"oneOf": [{ "const": "assistant" }, { "const": "system" }, { "const": "user" }],
						"title": "ChatCompletionInputMessageRole",
```

Wauplin replied:
Good idea yes

SBrandeis commented on packages/tasks/src/tasks/chat-completion/spec/input.json:
```diff
@@ -0,0 +1,54 @@
+				"properties": {
+					"role": {
+						"oneOf": [{ "const": "assistant" }, { "const": "system" }, { "const": "user" }],
+						"description": "The role of the messages author."
```
(nit)
```suggestion
						"description": "The role of the message author."
```

SBrandeis commented on packages/tasks/src/tasks/chat-completion/spec/input.json:
Ref: https://huggingface.github.io/text-generation-inference/#/Text%20Generation%20Inference/chat_completions

For docstrings / parameters

SBrandeis commented on packages/tasks/src/tasks/chat-completion/spec/input.json:
```diff
@@ -0,0 +1,54 @@
+			"type": "integer",
+			"description": "The random sampling seed."
+		},
+		"stop": {
+			"oneOf": [{ "type": "string" }, { "type": "array", "items": { "type": "string" } }],
+			"description": "Stop generating tokens if a stop token is generated."
+		},
```
Same comment - let's maybe add a `title`
```suggestion
		"stop": {
			"oneOf": [{ "type": "string" }, { "type": "array", "items": { "type": "string" } }],
            "title": "ChatCompletionInputStopReason",
			"description": "Stop generating tokens if a stop token is generated."
		},
```

SBrandeis replied:
Could even be moved to `common-definitions` if it makes sense to share the definition with `text-generation`

Wauplin replied:
I thought about it but Python-wise it makes things more difficult to handle. If that's fine with you, I'll prefer to keep them separated.

SBrandeis replied:
Sounds good!

SBrandeis commented on packages/tasks/src/tasks/chat-completion/spec/output_stream.json:
```diff
@@ -0,0 +1,48 @@
+						"title": "ChatCompletionStreamOutputDelta",
+						"description": "A chat completion delta generated by streamed model responses.",
+						"properties": {
+							"content": {
+								"type": "string",
+								"description": "The contents of the chunk message."
+							},
+							"role": {
+								"type": "string",
+								"description": "The role of the author of this message."
+							}
```
Those props should be required I think

SBrandeis commented on packages/tasks/src/tasks/chat-completion/spec/output_stream.json:
```diff
@@ -0,0 +1,48 @@
+						}
+					},
+					"finish_reason": {
+						"$ref": "#/$defs/FinishReason",
```
Interesting that this works while the definition is not in the file

If that works / is not too complex, I think it would be clearer to explicitly refer to the file where it is defined, like we do for classification outputs for ex:
https://github.com/huggingface/huggingface.js/blob/80698813868a11045616dc86ab2fdd1b34c493c4/packages/tasks/src/tasks/image-classification/spec/output.json#L7-L10

Wauplin replied:
Actually it works because all schemas (input, output, output_stream) are combined before generating code. But good idea yes, better to be explicit :+1:

SBrandeis commented on packages/tasks/src/tasks/chat-completion/spec/output.json:
```diff
@@ -0,0 +1,55 @@
+					"message": {
+						"type": "object",
+						"properties": {
+							"content": {
+								"type": "string",
+								"description": "The content of the chat completion message."
+							}
```
Missing `role` I think

SBrandeis reviewed (approved):
Super cool 🔥 
I like the way you specced the streaming output

I have a few minor comments, otherwise LGTM 👍

Wauplin commented:
Thanks for the great review @SBrandeis! I think I've addressed all of your comments by now :)
I fixed the lint CI but the test one still fails. It doesn't seem related to the PR though.

I'll this PR open a bit more in case someone else wants to review, otherwise I'll merge in the coming days.

Wauplin commented:
No complains so let's get this merged! :)
